README.md
# pomodorotimer (CLI Pomodoro Timer)
Console Pomodoro timer with playing the melody of the signal or with the 
generation of an end time signal. Displays a push notification on Linux and 
Mac OS X.

**Support**: [![Donate via PayPal](https://www.paypalobjects.com/webstatic/mktg/Logo/pp-logo-100px.png)](https://www.paypal.me/patsuckow) or [![Buy us a Cofee](https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png)](https://www.buymeacoffee.com/patsuckow)

# Pomodoro Technique:
Pomodoro Technique - [see](https://en.wikipedia.org/wiki/Pomodoro_Technique)

![plan](https://user-images.githubusercontent.com/12321741/72369481-459b5180-3711-11ea-9bb6-e6eed7a4a1e4.jpg)

We will use the traditional Pomodoro scheme:

![technique](https://user-images.githubusercontent.com/12321741/72372975-77fc7d00-3718-11ea-944b-69e426dc30b2.png)

### It is important:
**According to the principle of the Pomodoro technique - if you are
distracted while doing work, then you must interrupt the current time
and start a new Pomodoro. And this means that you need to write time in to
the database only after the specified time has passed and not earlier.**

**If you follow this principle, the Pomodoro Technique will be for you exactly 
the tool for which it was invented, namely, not to calculate the time of your 
work and breaks in work, but to make you concentrate on continuous and 
concentrated execution of works during one Pomodoro, i.e. within 25 minutes.**

---

We call the program from any directory in the console and set the countdown 
time:

![1](https://user-images.githubusercontent.com/12321741/72445421-bcdcee00-37c2-11ea-863f-9f3fc9c36a9f.gif)

When the time ends, a sound signal will be played, push notification will be 
displayed and it is proposed to enter a new countdown time. And also, it will 
display the time that is OUT, in the form of a colored time bar, where:
- red color - working time (25 minutes)
- green color - time for a short or long break (5, 15 or 30 minutes)

![2](https://user-images.githubusercontent.com/12321741/72445482-cfefbe00-37c2-11ea-9a0f-6fe319b13946.gif)

![3](https://user-images.githubusercontent.com/12321741/72445528-e433bb00-37c2-11ea-82bb-44fb220ba732.gif)

![4 2](https://user-images.githubusercontent.com/12321741/72445885-82278580-37c3-11ea-9edc-fc5dedd40276.gif)

![5 2](https://user-images.githubusercontent.com/12321741/72446142-0a0d8f80-37c4-11ea-94bf-fc789bfb426b.gif)

The type of push notification depends on your OS and the type of graphical 
shell used, for example:

![Peek 2019-12-30 22-35](https://user-images.githubusercontent.com/12321741/71597566-cde7f700-2b54-11ea-83a9-133cc737d32c.gif)

## How to install:

#### From PyPI:

    pip install pomodorotimer

#### From sources:

Alternatively you can install **pomodorotimer** from sources directory:

    git clone https://github.com/patsuckow/pomodorotimer
    cd pomodorotimer
    pip install -r requirements.txt
    pip install .
    cd ..
    rm -rf pomodorotimer

## How run **pomodorotimer**:
```
pomodoro
```

## Work with statistics Pomodoro`s:

Get statistics today:
```
pomodoro --statistic=today
```
We get something like this in the browser window:

![stat-2](https://user-images.githubusercontent.com/12321741/72270776-b9b4f700-3636-11ea-972b-e92d767beaad.jpg)

All-Time Statistics:
```
pomodoro --statistic=all
```
We get something like this in the browser window:
![all](https://user-images.githubusercontent.com/12321741/72447777-f879b700-37c6-11ea-9fda-edd623b8442f.png)

Delete (clear) statistics for today:
```
pomodoro --statistic=delete-today
```
Delete (clear) all statistics:
```
pomodoro --statistic=delete-all
```

## Requirements:
See in requirements.txt

**pomodorotimer** works with python 3.10.13 or higher.


## How uninstall **pomodorotimer** 😭:
```
pip uninstall pomodorotimer
```

## Licence:
![GNU GPL v 3 0](https://user-images.githubusercontent.com/12321741/67310082-c4636280-f505-11e9-83a7-d23e8037c54f.png)

## Authors:

**Alexey Patsukov 🇷🇺** - [GitHub profile](https://github.com/patsuckow)

### Submit issues

If you spotted something weird in application behavior or want to propose a 
feature you are welcome.

### Write code

If you are eager to participate in application development and to work on an 
existing issue (whether it should
be a bugfix or a feature implementation), fork, write code, and make a pull 
request right from the forked project page.

### Spread the word

If you have some tips and tricks or any other words that you think might be of 
interest for the others — publish it
wherever you find convenient.

Read also:
- https://francescocirillo.com
- https://habr.com/ru/post/446996/
- http://career-philol.ru/page/adaptive/id341091/blog/3189767/


CHANGELOG
=========
Changelog
=========

Versioning: 'major.minor.micro' (PEP 440)

Current versions
================

Version 0.2.5
-------------
- Added:
  - settings.py and config.ini
    Added a settings reading class and the settings file itself, containing the locales en_EN and ru_RU. 
    Settings will be taken from the file according to the locale of the user's PC.

Older versions
==============

Version 0.2.4
-------------
- Fixed: 
  - graph_statistic.py - Urgent correct, by stupidity, the broken right path to svg statistics files

Version 0.2.3
-------------
- Added:

- Changed: 
    - sound.py:
        - Changed sound.py so that it does not play as a separate process, but connects as a user module to pomodorotimer.py.
        For this, libraries were used as new dependencies: sunddevice, numpy and scipy.
        - Renamed methods for more information: default_sound->play_default_sound, wav_sound->play_wav_sound.
    - pomodorotimer.py:
        - The global constant UP_CLEAR_STR has been replaced with the property self.clear_up_str.
        - The global constant min_str has been replaced with the property self.scale_progress. And delete separator '|'.
        - Different minor code changes

- Fixed: 
    - sound.py: 
      - the correct path to the sound.py file
      - Used os.path.abspath and os.path.join for more reliable operation with the paths to the sound file.
    - pomodorotimer.py: 
      - Proper processing of interruption by pressing Ctr+C.
      - Cleaning the console, before entering the pomodoro timer, was transferred to a separate method.

- Updated: Almost all project files and README

Version 0.2.2
-------------

- Updated: Almost all project files and README
- Changed: Pomodoro will work with python 3.10.13 or higher. This is due to updates to the dependencies of connected libraries.
- Fixed: 
    - setup.py: The os module is removed. Added pathlib, which provides platform-independent operations over the paths. Fixed dependencies.
    - requirements.txt - Fixed dependencies. Removed the now unused numpy library.

Version 0.2.0
-------------

- Added: Saving to a local SQLite3 database file and displaying statistics in a browser window
- Updated: Almost all project files and README
- Changed: Pomodoro Cycle progress bar colors changed
- Fixed some bugs

Version 0.1.2 (31.12.2019)
--------------------------

- Fixed a bug


Version 0.1.1 (31.12.2019)
--------------------------

- Updated: README
- Changed:
    In order to hide the output of debugging information of the audio
    player, we will transfer the playback/generation code of the audio
    to a separate file and we will call this script in the background.
    Why so? - because standard output redirection does not work due to
    the specifics of ffmpeg it self.
    The process itself is needed to play the WAV file or the generated sound.
- Fixed some bugs


Version 0.1.0 (30.12.2019)
--------------------------

+ First release. Basic functionality.


pomodorotimer/sound.py
import os
import sounddevice as sd
import numpy as np
from scipy.io import wavfile

class Sound:
    def __init__(self):
        self.path = os.path.abspath(os.path.join(os.path.dirname(__file__), "tic-tic.wav"))
        if os.path.exists(self.path):
            self.play_wav_sound()
        else:
            self.play_default_sound()

    @staticmethod
    def play_default_sound() -> None:
        """
        Sound generation with a frequency of 432 Hz and a duration of 864 milliseconds
        """
        frequency = 432
        duration = 0.864
        # Timeline (44100 Hz - sampling rate)
        t = np.linspace(0, duration, int(44100 * duration), endpoint=False)
        # Generation of sinusoidal sound
        data = 0.5 * np.sin(2 * np.pi * frequency * t)
        # Playback
        sd.play(data, 44100)
        # Waiting for the playback to end
        sd.wait()
    
    def play_wav_sound(self) -> None:
        """Play wav file"""
        # Reading audio file
        fs, data = wavfile.read(self.path)
        # Playback
        sd.play(data, fs)
        # Waiting for the playback to end
        sd.wait()

if __name__ == "__main__":
    Sound()

pomodorotimer/settings.py
import os
import locale
import configparser


class Settings:
    def __init__(self):
        self.config = configparser.ConfigParser()
        self.load_config()

    def load_config(self):
        config_file = os.path.abspath(os.path.join(os.path.dirname(__file__), "config.ini"))
        # Check the existence of the configuration file
        if not os.path.isfile(config_file):
            raise FileNotFoundError(f"Configuration file '{config_file}'not found.")
        # Read the settings from the file using the ConfigParser() class method
        self.config.read(config_file)

    def get_locale(self) -> str:
        """Get a localized configuration from the settings file."""
        # Get the current system locale
        current_locale = locale.getlocale()[0]
        if current_locale in self.config:
            return current_locale
        else:
            return 'en_EN'
    
    def get_setting(self, option):
        """Get the string values of the requested option from the settings file."""
        # If the resulting line is '\\n', then first replacing it with '\n'
        return self.config.get(self.get_locale(), option).replace('\\n', '\n')

pomodorotimer/statistics_db.py
import os
import time
import sqlite3
import datetime


def create_table_statistic():
    sql = """
    CREATE TABLE IF NOT EXISTS 
    statistic(
        date TEXT,
        work INT,
        relaxation INT
    )
    """
    try:
        cursor.execute(sql)  # We execute the SQL query
    except sqlite3.DatabaseError as err:
        print("Error:", err)


def read_today_entry() -> tuple:
    try:
        sql = f"SELECT work, relaxation FROM statistic WHERE date = '{DATE}'"
        cursor.execute(sql)
    except sqlite3.DatabaseError as err:
        print("Error: ", err)

    work = 0
    relaxation = 0
    rows = cursor.fetchall()
    if rows:
        work = rows[0][0]
        relaxation = rows[0][1]

    return [DATE], [work], [relaxation]


def insert_entry(work: int, relaxation: int) -> None:
    sql = """INSERT INTO statistic(date, work, relaxation) VALUES (?, ?, ?)"""
    params = (DATE, work, relaxation)
    try:
        cursor.execute(sql, params)
    except sqlite3.DatabaseError as err:
        print("Error: ", err)
    else:
        conn.commit()  # Save changes to the database


def update_entry(work: int, relaxation: int) -> None:
    sql = f"""
    UPDATE statistic SET work={work}, relaxation={relaxation} 
    WHERE date='{DATE}'
    """
    try:
        cursor.execute(sql)
    except sqlite3.DatabaseError as err:
        print("Error: ", err)
    else:
        conn.commit()


def save_today_entry(work: int, relaxation: int) -> None:
    """
    In this case, you cannot use:
    'INSERT OR REPLACE INTO User VALUES(?,?,?)'
    """
    # First, check if there is a log entry for today
    _, work_db, relaxation_db = read_today_entry()
    if work_db == [0] and relaxation_db == [0]:
        # If the record does not exist, then add it
        insert_entry(work, relaxation)
    else:
        # Otherwise, update the record by ADDING new data
        update_entry(int(*work_db) + work, int(*relaxation_db) + relaxation)


def read_all_entry() -> tuple:
    try:
        cursor.execute("SELECT date, work, relaxation FROM statistic")
    except sqlite3.DatabaseError as err:
        print("Error: ", err)

    dates = []
    work = []
    relaxation = []
    rows = cursor.fetchall()
    if rows:
        for row in rows:
            dates.append(row[0])
            work.append(row[1])
            relaxation.append(row[2])

    return dates, work, relaxation


def delete_today_entry() -> None:
    try:
        cursor.execute(f"DELETE FROM statistic WHERE date='{DATE}'")
    except sqlite3.DatabaseError as err:
        print("Error: ", err)
    else:
        conn.commit()


def delete_all_entry() -> None:
    try:
        cursor.execute("DELETE FROM statistic")
    except sqlite3.DatabaseError as err:
        print("Error: ", err)
    else:
        conn.commit()


def close_db():
    cursor.close()
    conn.close()


# Create a connection object.
# If the database file is not found in the specified directory, then it will
# be created.
conn = sqlite3.connect(os.path.dirname(__file__) + "/statistics.db")

# According to the DB-API 2.0 specification, after creating a connection
# object, you must create a cursor object. All further requests must be
# made through this object. Creating a cursor object is done using the
# cursor() method.
# Create a cursor object (all database queries are executed through it).
cursor = conn.cursor()

# Today's date
DATE = str(datetime.datetime.fromtimestamp(time.time()).strftime("%Y-%m-%d"))

# Create table statistic if not exists
create_table_statistic()


pomodorotimer/graph_statistic.py
import os
import datetime
import calendar
import webbrowser
import matplotlib.pyplot as plt


class PeriodPomodoros:
    def __init__(
        self, dates: list, work: list, relaxation: list, name_svg: str
    ) -> None:
        # the width of the bars and size of the figure img
        if len(work) > 2:
            width = 0.4
            fig_size = (9.2, 7)
        else:
            width = 0.2
            fig_size = (7, 6)

        name_periods = []
        for i in dates:
            year, month, day = i.split("-")
            date = datetime.datetime(int(year), int(month), int(day)).weekday()
            name_periods.append(calendar.day_name[date] + "\n")

        if len(work) > 21:
            days = dates
        else:
            days = [k + j for k, j in zip(name_periods, dates)]

        x = range(len(days))  # the label locations
        fig, ax = plt.subplots(figsize=fig_size)
        rects_work = ax.bar([i - width / 2 for i in x], work, width, label="work")
        rects_relaxation = ax.bar([i + width / 2 for i in x], relaxation, width, label="relaxation")
        ax.set_title(f"Statistics Pomodoro`s for the last {len(work)} days:")
        ax.set_xticks(x)
        ax.set_xticklabels(days)
        ax.legend(bbox_to_anchor=(0.56, -0.2))
        ax.set_ylabel("Minutes")

        if len(work) > 7:
            for label in ax.xaxis.get_ticklabels():
                label.set_rotation(45 if len(work) <= 14 else 90)

        self.auto_label(rects_work, ax)
        self.auto_label(rects_relaxation, ax)
        fig.tight_layout()
        plt.savefig(name_svg, bbox_inches="tight")

    @staticmethod
    def auto_label(rects: list, ax) -> None:
        """
        Attach a text label above each bar in *rects*, displaying its height.
        """
        for rect in rects:
            height = rect.get_height()
            ax.annotate(
                "{}".format(height),
                xy=(rect.get_x() + rect.get_width() / 2, height),
                xytext=(0, 3),  # 3 points vertical offset
                textcoords="offset points",
                ha="center",
                va="bottom",
            )


class CountPomodorosTime:
    def __init__(self, work: list, relaxation: list, name_svg: str) -> None:
        work = sum(work)
        relaxation = sum(relaxation)

        fig, ax = plt.subplots(figsize=(9.2, 4.5))
        wedges, texts, autotexts = ax.pie(
            [work, relaxation],
            autopct=lambda pct: self.calc_pct(pct, [work, relaxation + 1]),
            textprops=dict(color="w"),
        )

        ax.legend(
            wedges,
            [
                "work: {:.2f} hours".format(work / 60),
                "relaxation: {:.2f} hours".format(relaxation / 60),
            ],
            title="Count Pomodoro`s time:\n",
            loc="center left",
            bbox_to_anchor=(0.92, 0, 0.5, 1),
        )

        plt.savefig(name_svg, bbox_inches="tight")

    @staticmethod
    def calc_pct(pct: int, values: list) -> str:
        return "{} min\n{:.2f}%".format(int(pct / 100 * sum(values)), pct)


def graph(dates: list, work: list, relaxation: list) -> None:
    path = os.path.abspath(os.path.dirname(__file__))

    # If there are old statistics image files, then delete them
    svg_1 = path + "stat-1.svg"
    if os.path.exists(svg_1):
        os.remove(svg_1)
    svg_2 = path + "stat-2.svg"
    if os.path.exists(svg_2):
        os.remove(svg_2)

    # If statistics for the requested period exists in the database, then
    # display it
    if work:
        if 1 < len(work) <= 31:
            PeriodPomodoros(
                dates=dates,
                work=work,
                relaxation=relaxation,
                name_svg=path + "stat-1.svg",
            )

        CountPomodorosTime(
            work=work, relaxation=relaxation, name_svg=path + "stat-2.svg"
        )

    webbrowser.open("file://" + path + "/stats.html")

pomodorotimer/pomodorotimer.py
import sys
import time
import argparse
import subprocess
from collections import deque
from .graph_statistic import graph
from .sound import Sound as play_sound
from .statistics_db import (
    save_today_entry,
    read_all_entry,
    read_today_entry,
    delete_today_entry,
    delete_all_entry,
    close_db,
)
from .settings import Settings


try:
    assert sys.version_info >= (3, 10, 13)
except Exception:
    raise SystemExit("Pomodoro Timer works with Python 3.10.13 or higher.")


class PomodoroTimer:
    """
    Created by Alexey Patsukov (patsuckow@yandex.ru) 🇷🇺
    CLI Pomodoro Timer - https://github.com/patsuckow/pomodorotimer
    """
    def __init__(self) -> None:
        # Get settings from an ini file
        setting = Settings()
        self.mess_input = f"{setting.get_setting('mess_input')} "
        self.mess_stop = setting.get_setting('mess_stop')
        self.mess_cycle_over = setting.get_setting('mess_cycle_over')
        self.mess_times_up = setting.get_setting('mess_times_up')

        self.scale_progress = ""
        self.clear_up_str = f"\x1b[A{79*' '}\r"
        self.stack_watch = deque(
            list("🕛🕧🕐🕜🕑🕝🕒🕞🕓🕟🕔🕠🕕🕡🕖🕢🕗🕣🕘🕤🕙🕥🕚🕦")
        )

        self.process_argument()
        self.pomodoro_setup()

    def process_argument(self) -> None:
        """
        Check if the correct one for working with statistics is passed;
        And depending on this, we make requests to the database or not.
        """
        argument = self.arg()
        if argument in ["today", "all", "delete-today", "delete-all"]:
            if argument == "today":
                dates, work, relaxation = read_today_entry()
                graph(dates, work, relaxation)
            elif argument == "all":
                dates, work, relaxation = read_all_entry()
                graph(dates, work, relaxation)
            elif argument == "delete-today":
                delete_today_entry()
            elif argument == "delete-all":
                delete_all_entry()

            close_db()
            self.stop_pomodoro()

    def pomodoro_setup(self) -> None:
        self.clear_console()

        try:
            while True:
                try:
                    self.show_console_cursor()
                    self.min = int(input(self.mess_input))
                except ValueError:
                    self.write_flush(self.clear_up_str)
                    continue

                self.write_flush(self.clear_up_str)

                if self.min in [5, 15, 25, 30]:
                    self.run_pomodoro()
                elif self.min == 0:
                    self.write_flush(self.clear_up_str)
                    if self.scale_progress != "":
                        print(self.scale_progress + self.mess_stop)
                    self.stop_pomodoro()
        except KeyboardInterrupt:
            # interrupt processing (Ctrl + C)
            print(self.mess_stop)
            self.stop_pomodoro()

    @staticmethod
    def clear_console() -> None:
        if sys.platform in ("linux", "darwin"):
            subprocess.call("clear", shell=True)
        elif sys.platform in ("win32", "cygwin"):
            subprocess.call("cls", shell=True)

    @staticmethod
    def arg() -> str:
        """Get the value of the argument --statistic passed to the script"""
        parser = argparse.ArgumentParser()
        parser.add_argument(
            "--statistic",
            type=str,
            default="",
            help="provide a string: 'today', 'all', 'delete-today', "
            "'delete-all' (default: '')",
        )

        return parser.parse_args().statistic

    def progress_scale(self) -> str:
        """
        Colorize the timer scale for pomodoro in a different color, depending on
        the set time.
        """
        match self.min:
            case 5:
                self.scale_progress += f"\x1b[42m\x1b[1m {self.min} \x1b[0m"
            case 15:
                self.scale_progress += f"\x1b[42m\x1b[1m  {self.min}  \x1b[0m"
            case 25:
                self.scale_progress += f"\x1b[41m\x1b[1m   {self.min}   \x1b[0m"
            case 30:
                self.scale_progress += f"\x1b[42m\x1b[1m    {self.min}    \x1b[0m"

        self.write_flush(self.clear_up_str)

        return self.scale_progress

    def run_reverse_timer(self) -> None:
        """Countdown"""
        self.write_flush("\n" + self.clear_up_str)
        
        for i in range(self.min * 60, 0, -1):
            self.write_flush(f"\r{self.animated_clock()}{self.lead_time(i)} ")
            # The code execution delay (by 1 second) allows you to change the display
            #  of a string with a timer every second, t.e. this is the timer itself.
            # (For yourself: If you change the code, then for quick checking of 
            # performance, comment on this line and the code will be executed instantly)
            time.sleep(1)

        self.write_flush(f"\r{35*' '}\r")

    @staticmethod
    def write_flush(string: str) -> None:
        """
        The sys.stdout.flush() method is needed to clear the buffer output of
        the standard output stream. T.e. all data that has been delayed in the
        output buffer will be sent to the output device (for example, to the console)
        immediately. This is useful when you need to make sure that all the data was
        printed before the code continues to execute.
        """
        sys.stdout.write(string)
        sys.stdout.flush()

    @staticmethod
    def hide_console_cursor() -> None:
        """Hide console cursor: '\x1b[?25l'
        https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_codes
        ('\033' same as '\x1b')
        """
        sys.stdout.write("\x1b[?25l")

    @staticmethod
    def show_console_cursor() -> None:
        """Show console cursor: '\x1b[?25h'"""
        sys.stdout.write("\x1b[?25h")

    def lead_time(self, sec: int) -> str:
        """Elapsed Time Calculation"""
        minutes = int(sec // 60)
        seconds = round(sec % 60, 3)
        if seconds // 10 == 0:
            seconds = "0" + str(seconds)
        if self.min == 25:
            return f" Pomodoro timer: \x1b[31m\x1b[1m{minutes}:{seconds}\x1b[0m"
        elif self.min in [5, 15, 30]:
            return f" Pomodoro timer: \x1b[32m\x1b[1m{minutes}:{seconds}\x1b[0m"

    def animated_clock(self) -> str:
        """
        To extract the first element from the list and place it at the end of
        the list, for greater performance, we will use the stack instead of
        the standard list methods .append() and .pop().
        https://docs.python.org/3/library/collections.html?#collections.deque
        """
        clock = self.stack_watch[0]
        self.stack_watch.appendleft(self.stack_watch.pop())

        return clock

    def save_statistics(self) -> None:
        """INSERT / UPDATE log statistic in DB

        According to the principle of the Pomodoro technique - if you are
        distracted while doing work, then you must interrupt the current time
        and start a new Pomodoro. And this means that you need to write time
        in to the database only after the specified time has passed and not
        earlier.

        If you follow this principle, the Pomodoro Technique will be for you
        exactly the tool for which it was invented, namely, not to calculate
        the time of your work and breaks in work, but to make you concentrate
        on continuous and concentrated execution of works during one Pomodoro,
        i.e. within 25 minutes.
        """
        work = 0
        relaxation = 0
        if self.min in [5, 15, 30]:
            relaxation = self.min
        else:
            work = self.min

        save_today_entry(work, relaxation)

    def show_pop_up_notice(self):
        """
        Show pop-up notification on Linux/macOS desktop.
        (Windows OS does not support the display of pop-up notifications.)
        """
        title = "⏰ Pomodoro: "
        info = f"{self.min} {self.mess_times_up}"

        if sys.platform.startswith("linux"):
            command = ["notify-send", title, info]
        elif sys.platform == "darwin":  # macOS
            command = [
                "osascript",
                "-e",
                f'display notification "{info}" with title "{title}"',
            ]
        else:
            return  # Do nothing if not on supported platform

        subprocess.run(command, check=False)

    def check_one_pomodoro_cycle(self) -> None:
        """
        Checking the end of the pomodoro cycle
        """
        if self.min in [15, 30]:
            return print(self.mess_cycle_over), self.stop_pomodoro()

    def run_pomodoro(self) -> None:
        """Run pomodoro cicle"""
        # Нide the cursor display from the console
        self.hide_console_cursor()
        # Start the countdown timer
        self.run_reverse_timer()
        # Display (update display) the pomodoro progress bar
        print(self.progress_scale())
        # At the end of the timer, play the sound signal
        self.show_pop_up_notice()
        # Save statistics in the database
        play_sound()
        # Show pop-up notice notification
        self.save_statistics()
        # Check if the latter introduced 15 or 30, then we complete 
        # the pomodoro execution cycle
        self.check_one_pomodoro_cycle()
    
    def stop_pomodoro(self) -> None:
        self.show_console_cursor()
        exit(0)